pubsublite-spark-sql-streaming/src/main/java/com/google/cloud/pubsublite/spark/PslMicroBatchReader.java

```diff
@@ -0,0 +1,154 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.pubsublite.spark;
+
+import com.google.cloud.pubsublite.Partition;
+import com.google.cloud.pubsublite.SubscriptionPath;
+import com.google.cloud.pubsublite.TopicPath;
+import com.google.cloud.pubsublite.cloudpubsub.FlowControlSettings;
+import com.google.cloud.pubsublite.internal.CheckedApiException;
+import com.google.cloud.pubsublite.internal.CursorClient;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.concurrent.ExecutionException;
+import java.util.stream.Collectors;
+import org.apache.spark.sql.catalyst.InternalRow;
+import org.apache.spark.sql.sources.v2.reader.InputPartition;
+import org.apache.spark.sql.sources.v2.reader.streaming.MicroBatchReader;
+import org.apache.spark.sql.sources.v2.reader.streaming.Offset;
+import org.apache.spark.sql.types.StructType;
+
+public class PslMicroBatchReader implements MicroBatchReader {
+
+  private final CursorClient cursorClient;
+  private final HeadOffsetReader headOffsetReader;
+  private final MultiPartitionCommitter committer;
+  private final SubscriptionPath subscriptionPath;
+  private final TopicPath topicPath;
+  private final FlowControlSettings flowControlSettings;
+  private final long topicPartitionCount;
+  private SparkSourceOffset startOffset;
+  private SparkSourceOffset endOffset;
+
+  public PslMicroBatchReader(
+      CursorClient cursorClient,
+      HeadOffsetReader headOffsetReader,
+      MultiPartitionCommitter committer,
+      SubscriptionPath subscriptionPath,
+      TopicPath topicPath,
+      FlowControlSettings flowControlSettings,
+      long topicPartitionCount) {
+    this.cursorClient = cursorClient;
+    this.headOffsetReader = headOffsetReader;
+    this.committer = committer;
+    this.subscriptionPath = subscriptionPath;
+    this.topicPath = topicPath;
+    this.flowControlSettings = flowControlSettings;
+    this.topicPartitionCount = topicPartitionCount;
+  }
+
+  @Override
+  public void setOffsetRange(Optional<Offset> start, Optional<Offset> end) {
+    if (start.isPresent()) {
+      assert SparkSourceOffset.class.isAssignableFrom(start.get().getClass())
+          : "start offset is not assignable to PslSourceOffset.";
+      startOffset = (SparkSourceOffset) start.get();
+    } else {
+      try {
+        Map<Partition, com.google.cloud.pubsublite.Offset> pslSourceOffsetMap = new HashMap<>();
+        for (int i = 0; i < topicPartitionCount; i++) {
+          pslSourceOffsetMap.put(Partition.of(i), com.google.cloud.pubsublite.Offset.of(0));
+        }
+        cursorClient
```

Comment:
this method will not return a value if the offset for a partition has never been set.

Reply:
Yea, I noticed that, that's why I have just above to get partition count and supply default value.